### ISSUE TITLE
Unexpected lock when two CLI clients are active

### ISSUE THREAD

**Issue by pkolaczk** [scalacenter/bloop#982]
I can't run a test or do anything with bloop from another terminal when the background compilation is waiting for file changes. 

I get:

```Waiting on external CLI client to release lock on this build...```

Why is it needed? 
Why can't we release the lock during the time it is *not* compiling and just waiting?

This seriously limits usefullness of background compilation mode. 
We care about every 0.1 second, but this lock is adding 2 more manual steps to stop watching and then to restart watching after running a test.

(BTW: I know I can use watching to also run tests, however I may have a larger set of tests that is slow and consuming lot of resources, so I *don't* want my tests running when I'm working with the code, until I finish).

**Comment by jvican** [comment]
At the moment, this use case is not supported on purpose. You can have multiple BSP clients per build but you can only have one CLI. The reason for this is because a CLI has a stable set of compile directories and stable directories can't be properly isolated across concurrent clients.

This is likely to change with some changes I'll work on soon, where a CLI won't use nailgun anymore and instead will act as a BSP client. However, I need to experiment with that a bit because that solution also has some drawbacks. There are other solutions we could implement to support this behavior too, let's discuss those when I provide feedback after running this experiment.
  
Note how asking to run a test from a BSP client such as Metals or IntelliJ works (at least when IntelliJ fixes the issue with testing).

**Comment by tindzk** [comment]
I hit the same problem in several projects. It is a common use case in client-server applications.

In previous Bloop versions, I could open two terminals with:

```shell
bloop link client --watch   # Scala.js frontend
bloop run server --watch    # Scala JVM backend
```

But with Bloop v1.3.2, the second command will wait for the lock to be released. The lock is actually not needed since `client` and `server` do not have any overlapping `.class` files.

If BSP is not affected by this limitation, I could adapt Seed to use a BSP connection instead of the Nailgun client. I had a look at the latest protocol definition, but it does not seem to be capable of linking modules yet.

**Comment by jvican** [comment]
@tindzk All build tool integrations should exist over BSP instead of CLI. Once a BSP integration is in place, you'll be able to run two seed commands where each of it uses a different part of the build graph.

**Comment by jvican** [comment]
I'm closing this ticket because this is a design decision I'm not planning to revisit any time soon. I might experiment with productionizing a prototype to make the CLI a bsp client down the road, but for now it's not a priority. Thanks for reporting, I think it's great there's some trace of this behavior in the issue tracker as well as a reiteration that build tool integrations should always prefer BSP-based instead of CLI-based integrations.

**Comment by jvican** [comment]
Note for casual readers: the workaround to this problem is to run tests/application from other Bloop clients than the CLI client. For example, you can run your application once from your CLI and from your editor concurrently (IntelliJ/VS Code/vim, etc). The problem discussed here only affects concurrent CLI clients.

**Comment by jypma** [comment]
Interesting. Why would CLI clients require a different kind of locking from editor clients?

**Comment by jvican** [comment]
@jypma Because BSP clients are isolated (every BSP client has its own unique set of classes directories granted upon their connection) while the CLI classes directories are stable (at least for now). You can read more about this in my Scala World talk https://www.youtube.com/watch?v=PgQyCnqm4QQ

**Comment by jvican** [comment]
Hi all! I found a sensible way to guarantee the isolation of CLI clients so the next release will remove the CLI lock and allow users to use as many concurrent CLI clients as they want.
